### ISSUE TITLE
Can't find installed candidates in fish

### ISSUE THREAD

**Issue by gabrielfeo** [deather/omf-sdk#6]
I have no idea if this could be related to omf-sdk, but I'm at a loss here. I can find candidates in zsh, but in fish sdkman is unaware of them. Same user.

The shell with syntax highlighting is fish:

<img width="1184" alt="Screen Shot 2021-02-04 at 16 52 52" src="https://user-images.githubusercontent.com/25730717/106948271-3fa45400-670a-11eb-89d4-3e8f6214fec2.png">

<img width="842" alt="Screen Shot 2021-02-04 at 16 57 18" src="https://user-images.githubusercontent.com/25730717/106948079-0c61c500-670a-11eb-905b-58b6c2529ea5.png">

<img width="286" alt="Screen Shot 2021-02-04 at 16 54 32" src="https://user-images.githubusercontent.com/25730717/106947826-a8d79780-6709-11eb-9dc8-874af837df74.png">

I can't link candidates in fish either. It says "Invalid path" even though it's not malformed and sdk in zsh accepted it:

<img width="1181" alt="Screen Shot 2021-02-04 at 17 03 33" src="https://user-images.githubusercontent.com/25730717/106948964-24861400-670b-11eb-97b4-881abc453d73.png">

**Comment by gabrielfeo** [comment]
**TL; DR:** SDKMAN_DIR must be defined in bash. This is what my `sdk.fish` does now:

```sh
bash -c "SDKMAN_DIR=$sdkman_prefix && source $sdkman_prefix/bin/sdkman-init.sh && sdk $argv"
```

<img width="1375" alt="Screen Shot 2021-02-04 at 20 42 36" src="https://user-images.githubusercontent.com/25730717/106969407-85244980-6729-11eb-9b40-d1778d444f06.png">

<img width="1260" alt="Screen Shot 2021-02-04 at 20 57 41" src="https://user-images.githubusercontent.com/25730717/106970460-a71ecb80-672b-11eb-8684-ef51d692e7d4.png">

@deather, would you be open to a PR with this change?

[This comment](https://github.com/sdkman/sdkman-cli/issues/431#issuecomment-656861492) gave me the clue. Since the bash omf-sdk starts ([sdk.fish](https://github.com/deather/omf-sdk/blob/37cbf40def97f5ede1f0805e8bd77898ba70ee61/functions/sdk.fish)) is not a login shell, it doesn't run any of the profile scripts which would set `SDKMAN_DIR`. It doesn't have any of the variables I set in fish either.

I don't really understand why sdkman behaves like this if `SDKMAN_DIR` isn't set though, since what I have set is actually the default location, or why this solved the "Invalid path" problem.

**Comment by deather** [comment]
Hi @gabrielfeo,

Thanks to take the time to post the bug. Normally that is the `init.fish` which set the `SDKMAN_DIR` automatically. The only thing that I see why this variable is not set, it is due to `fenv` is missing.

- Can you check first if `fenv` is installed with this command `type fenv` or checking if the folder `.local/share/omf/pkg/` exists ?
- Can you post the result of `env | grep SDK` in fish and `bash -c "env" | grep SDK` ?

We don't use the bash as login shell (-l option if I understood). I want to understand where is the problem before create the PR.

**Comment by gabrielfeo** [comment]
Thank you for your help. SDKMAN_DIR is indeed defined:

<img width="682" alt="Screen Shot 2021-02-05 at 19 21 53" src="https://user-images.githubusercontent.com/25730717/107095671-faeeea80-67e7-11eb-9096-a2e8f2dd2af6.png">
<img width="927" alt="Screen Shot 2021-02-05 at 19 37 13" src="https://user-images.githubusercontent.com/25730717/107096476-9df43400-67e9-11eb-8f55-e7fc44cb26ba.png">
<img width="927" alt="Screen Shot 2021-02-05 at 19 37 26" src="https://user-images.githubusercontent.com/25730717/107096477-a0568e00-67e9-11eb-8dc3-376e0e1be388.png">

I have a lot to learn about shells. I'd appreciate it if you could confirm if I got this right.
 1. `SDKMAN_DIR` **will** get passed down to `bash` because you _export_ it
 2. `fenv` makes sure when sdkman changes JAVA_HOME from within `bash`, the change gets passed up to `fish`

Back to the problem, I now have no clue as to why re-defining the variable solves the issue:

<img width="927" alt="Screen Shot 2021-02-05 at 19 35 56" src="https://user-images.githubusercontent.com/25730717/107096362-5f5e7980-67e9-11eb-9b18-b73556ed2d59.png">

**Comment by deather** [comment]
Your understanding is good about fenv, and export. If you read `man set` you will find more information about the option. The trick you purposed is to define `SDKMAN_DIR` for one session of bash.

The only thing we don't check is the `.bash_profile` and the `.bashrc`. Wen you use `bash` it loads automatically some files like these two.

- Can you try `bash --norc --noprofile -c "source $sdkman_prefix/bin/sdkman-init.sh && sdk l java | grep '>>>'"` ?
- Can you paste the content of the `.bashrc` file ?

**Comment by gabrielfeo** [comment]
Here are the outputs of those commands. I don't have a `.bashrc`. I do have a `.bash_profile`, but it's probably irrelevant because it's not run in a non-login shell, according to [this doc](https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html).

<img width="1088" alt="Screen Shot 2021-02-08 at 17 43 28" src="https://user-images.githubusercontent.com/25730717/107278928-29add080-6a35-11eb-8ae4-eb78e7461f84.png">

I just setup fish, omf and omf-sdk on another Mac to test this, and I had no issues. It's probably something local, maybe due to my home directory on this machine being in a non-standard location.

**Comment by deather** [comment]
When you said your home directory is not standard, you have redefined it with these kind of command https://stackoverflow.com/a/26427748/1595433 ? or you just change the `$HOME` variable ?

Can you give me the output of `bash -c "echo $HOME"`, `bash -c "echo ~"` and `bash -c "ls -la ~ | grep .sdkman"` ?

**Comment by gabrielfeo** [comment]
I moved my home directory using macOS's System Preferences:

<img width="780" alt="Screen Shot 2021-02-15 at 20 26 40" src="https://user-images.githubusercontent.com/25730717/108001754-1f0eb080-6fcc-11eb-9619-299e2d3bee60.png">

<img width="682" alt="Screen Shot 2021-02-15 at 20 25 08" src="https://user-images.githubusercontent.com/25730717/108001687-e7a00400-6fcb-11eb-8b32-74efa8b47d00.png">
